lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from django.db.utils import IntegrityError
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from base64 import b64encode
+from collections import namedtuple
+from hashlib import sha256
+import json
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+BlockRecord = namedtuple('BlockRecord', ['locator', 'weight', 'max_score'])
+"""
+A namedtuple used to serialize information about a block at the time it was used in grade calculation.
+"""
+
+
+def blockrecord_list_to_json(blocks):
+    """
+    Return a JSON-serialized version of the list of block records.
+    """
+    return json.dumps([block._asdict() for block in blocks], separators=(',', ':'), sort_keys=True)
+
+def blockrecord_json_to_hash(block_json):
+    """
+    Return a hashed version of the list of block records.
+
+    This currently hashes using sha256, and returns a b64 encoded version
+    of the binary digest.
+    """
+    return b64encode(sha256(block_json).digest())
```

Comment:
Where do we plan to keep the version number if we ever decide to change the serialization protocol?